Prompt API: Add overview for Prompt API (#43848)
mdn/content: files/en-us/web/api/languagemodel/promptstreaming/index.md

---
title: "LanguageModel: promptStreaming() method"
short-title: promptStreaming()
slug: Web/API/LanguageModel/promptStreaming
page-type: web-api-instance-method
spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{APIRef("Prompt API")}}{{SecureContext_Header}}

> The **`promptStreaming()`** method of the {{domxref("LanguageModel")}} interface sends input to the language model and returns a {{domxref("ReadableStream")}} that delivers the model's response incrementally as it is generated. This is useful for displaying responses to users in real time, for long outputs, or for any scenario where low perceived latency matters. Consume the stream using `for await...of` or by attaching a reader via {{domxref("ReadableStream.getReader()")}}.

Reviewer: ```suggestion
The **`promptStreaming()`** method of the {{domxref("LanguageModel")}} interface sends input to the language model and returns a {{domxref("ReadableStream")}} that delivers the model's response incrementally as it is generated.

This is useful for displaying responses to users in real time, for long outputs, or for any scenario where low perceived latency matters. Consume the stream using `for await...of` or by attaching a reader via {{domxref("ReadableStream.getReader()")}}.
```

## Syntax

```js-nolint
promptStreaming(input)
promptStreaming(input, options)
```

### Parameters

- `input`
  - : The prompt to send to the model. This is a `LanguageModelPrompt`, which is either:
    - A string — Shorthand for a single user message. For example: `[{ role: "user", content: [{ type: "text", value: input }] }]`.

>     - A sequence representing a single message in a conversation with a language model. Options include:

Reviewer: Sequence.

Author: Next commit.

      - `role`
        - : A string indicating who sent the message. Must be one of:
          - `"system"`
            - : A system-level instruction that guides the model's overall behavior. Note that {{domxref("LanguageModel.prompt()", "prompt()")}}, {{domxref("LanguageModel.promptStreaming()", "promptStreaming()")}}, {{domxref("LanguageModel.append()", "append()")}} throw a `"NotSupportedError"` `DOMException` if a message with `role: "system"` is passed to them; system messages are only allowed in `initialPrompts`.
          - `"user"`
            - : message from the user.
          - `"assistant"`
            - : A message from the model (used for few-shot examples or continued dialogue).
- `options` {{optional_inline}}
  - : Options for streaming a prompt. Options include:
    - `responseConstraint`

>       - : Constraints on the format of the model's output. When provided and `omitResponseConstraintInput` is `false`, any implementation-defined constraint-description message is included in the measurement.

Reviewer: Again, type of this, and format. In concepts you may already (but either way should) explain that you can apply constraints.

Author: As before.

---
title: "LanguageModel: promptStreaming() method"
short-title: promptStreaming()
slug: Web/API/LanguageModel/promptStreaming
page-type: web-api-instance-method
browser-compat: api.LanguageModel.promptStreaming
---

{{APIRef("Prompt API")}}{{SecureContext_Header}}

The **`promptStreaming()`** method of the {{domxref("LanguageModel")}} interface sends input to the language model and returns a {{domxref("ReadableStream")}} that delivers the model's response incrementally as it is generated.

> This is useful for displaying responses to users in real time, for long outputs, or for any scenario where perceived latency should be minimized. Consume the stream using `for await...of` or by attaching a reader via {{domxref("ReadableStream.getReader()")}}.

Reviewer: ```suggestion
This is useful for displaying responses to users incrementally for outputs that take a long time to complete, or for any scenario where perceived latency should be minimized. Consume the stream using `for await...of` or by attaching a reader via {{domxref("ReadableStream.getReader()")}}.
```

Each call to `promptStreaming()` adds to the session's context. To branch from a given state without affecting the original session, call {{domxref("LanguageModel.clone()")}}.

## Examples

### Streaming a response to the page

This example writes out chunks from a `promptStreaming()` call's {{domxref("ReadableStream")}} as they arrive.

```js
const session = await LanguageModel.create();
const output = document.querySelector("#output");

const stream = session.promptStreaming("Write a short poem about the ocean.");

for await (const chunk of stream) {
  output.textContent += chunk;
}
```

### Streaming with an abort signal

This example shows how to use an {{domxref("AbortController")}} with `promptStreaming()`.

```js
const controller = new AbortController();
document
  .querySelector("#stop")
  .addEventListener("click", () => controller.abort());

const stream = session.promptStreaming("Tell me a long story.", {
  signal: controller.signal,
});

try {
  for await (const chunk of stream) {

>     console.log(chunk);

Reviewer: FWIW and in other cases where you show streaming, console logging is not great because you end up with so many small chunks. If you want someone to copy/paste this into dev tools to run then fine, but to show a more natural usage I'd suggest maybe writing to a log_append() function or similar.

I.e. not a blocker, just a thought from having tried copying the example in the guide.

Reviewer 2: I...didn't actually have an example in any of my guides. To remedy your point and fix this, I:

- Updated instances of `console.log(chunk)` to `output.textContent += chunk;`, as it is already used ina. few places, and I thought it was simple enough to infer what is going on.
- Added a "Complete streaming example" section at the bottom of the "Using..." guide.

Reviewer: There was an example somewhere - maybe in reference. But that sounds like a good solution.

Reviewer 2: It's nice to have a full, working example available. It looks very pleasing when the chunks are written to the output gradually.
